pkg/process/events/listener_linux_test.go

```diff
@@ -0,0 +1,188 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"io"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/DataDog/datadog-agent/pkg/process/events/model"
+	"github.com/DataDog/datadog-agent/pkg/security/api"
+	"github.com/DataDog/datadog-agent/pkg/security/api/mocks"
+)
+
+// assertProcessEvents compares two ProcessEvents. Two events can't be compared using directly assert.Equal
+// due to the embedded time fields
+func assertProcessEvents(t *testing.T, expected, actual *model.ProcessEvent) {
+	assert.Equal(t, expected.EventType, actual.EventType)
+	assert.WithinDuration(t, expected.CollectionTime, actual.CollectionTime, 0)
+	assert.Equal(t, expected.Pid, actual.Pid)
+	assert.Equal(t, expected.Ppid, actual.Ppid)
+	assert.Equal(t, expected.UID, actual.UID)
+	assert.Equal(t, expected.GID, actual.GID)
+	assert.Equal(t, expected.Username, actual.Username)
+	assert.Equal(t, expected.Group, actual.Group)
+	assert.Equal(t, expected.Exe, actual.Exe)
+	assert.Equal(t, expected.Cmdline, actual.Cmdline)
+	assert.WithinDuration(t, expected.ForkTime, actual.ForkTime, 0)
+	assert.WithinDuration(t, expected.ExecTime, actual.ExecTime, 0)
+	assert.WithinDuration(t, expected.ExitTime, actual.ExitTime, 0)
+}
+
+// TestProcessEventFiltering asserts that sysProbeListener collects only expected events and drops everything else
+func TestProcessEventFiltering(t *testing.T) {
+	rawEvents := make([]*model.ProcessMonitoringEvent, 0, 3)
+	handlers := make([]EventHandler, 0, 3)
+
+	// The listener should drop unexpected events and not call the EventHandler for it
+	rawEvents = append(rawEvents, model.NewProcessMonitoringEvent(model.Fork, time.Now(), 23, "/usr/bin/ls", []string{"ls", "-lah"}))
+
+	// Verify that expected events are correctly consumed
+	rawEvents = append(rawEvents, model.NewProcessMonitoringEvent(model.Exec, time.Now(), 23, "/usr/bin/ls", []string{"ls", "-lah"}))
+	handlers = append(handlers, func(e *model.ProcessEvent) {
+		require.Equal(t, model.Exec, e.EventType)
+		require.Equal(t, uint32(23), e.Pid)
+	})
+
+	rawEvents = append(rawEvents, model.NewProcessMonitoringEvent(model.Exit, time.Now(), 23, "/usr/bin/ls", []string{"ls", "-lah"}))
+	handlers = append(handlers, func(e *model.ProcessEvent) {
+		require.Equal(t, model.Exit, e.EventType)
+		require.Equal(t, uint32(23), e.Pid)
+	})
+
+	// To avoid race conditions, all handlers should be assigned during the creation of SysProbeListener
+	i := 0
+	handler := func(e *model.ProcessEvent) {
+		handlers[i](e)
+		i++
+	}
+
+	l, err := newSysProbeListener(nil, nil, handler)
+	require.NoError(t, err)
+
+	for _, e := range rawEvents {
+		data, err := e.MarshalMsg(nil)
+		require.NoError(t, err)
+		l.consumeData(data)
+	}
```

Comment:
We should add an assertion here to make sure all handlers are called and not dropped.

Reply:
Done in https://github.com/DataDog/datadog-agent/pull/12099/commits/3483bcc299ebda145257a11d10995abdb11148af and https://github.com/DataDog/datadog-agent/pull/12099/commits/564beab9caa3c59be70329e66e06562877c63163.